File: LeetCode/209.Minimum_Size_Subarray_Sum/Mario/209.Minimum_Size_Subarray_Sum.py
#!/usr/bin/env python3


class Solution:
    def minSubArrayLen(self, target: int, nums: List[int]) -> int:
        # sliding window
        n = len(nums)
        res = math.inf
        l, cur = 0, 0

        for r, v in enumerate(nums):
            cur += v
            while l <= r and cur >= target:
                res = min(res, r - l + 1)
                cur -= nums[l]
                l += 1

        return 0 if res == math.inf else res

File: LeetCode/209.Minimum_Size_Subarray_Sum/209.Minimum_Size_Subarray_Sum.java
class Solution {
    public int minSubArrayLen(int target, int[] nums) {
        int i = 0, j = 0;
        int n = nums.length;
        int minLen = n + 1;
        int sum = nums[0];
        while (j < n) {
            if (sum >= target) {
                if (j - i + 1 < minLen)
                    minLen = j - i + 1;
                sum -= nums[i];
                i += 1;
            } else {
                j += 1;
                if (j < n)
                    sum += nums[j];
            }
        }
        return minLen == n + 1 ? 0 : minLen;
    }
}